src/main/java/pl/edu/mimuw/Bank.java

```diff
@@ -1,25 +1,134 @@
-  }
+    public Bank(double depositInterestRate, double loanInterestRate) {
+        this.depositInterestRate = depositInterestRate;
+        this.loanInterestRate = loanInterestRate;
+    }
 
-  public void addClient(BankClient client) {
-    throw new IllegalStateException("TODO");
-  }
+    private String assignActionNumber() {
+        return String.format("%04d", actionCount++);
+    }
 
-  public BankClient getClient(int clientIndex) {
-    final var clients = getAllClients();
-    if (clientIndex >= clients.size()) throw new IllegalArgumentException("Invalid client index");
-    return clients.get(clientIndex);
-  }
+    private String assignClientNumber() {
+        return String.format("%04d", clientCount++);
+    }
+
+    public void collectInstallments() {
+        for (var client : allClients)
+            for (var action : client.getActionHistory())
+                action.collect(this, client);
+    }
+
+    public void advanceCurrentDate(int days) {
+        Calendar date = Calendar.getInstance();
+
+        date.setTime(this.today);
+        date.add(Calendar.DATE, days);
+
+        this.today = Timestamp.from(date.toInstant());
+    }
+
+    public Timestamp getCurrentDate() {
+        return this.today;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder result = new StringBuilder();
+        result.append("BANK STATS\nList of clients:\n");
+
+        for (var client : allClients) {
+            result.append(client.getCredentials()).
+                    append(" acc bal: ").append(client.getBalance())
+                    .append(", number " + "of actions: ")
+                    .append(client.getActionHistory().size()).append("\n");
+            for (var action : client.getActionHistory())
+                result.append(" * ").append(action.toString()).append("\n");
+        }
+
+        result.append("Interest rates: \nDeposit: ")
+                .append(this.depositInterestRate)
+                .append(", loan: ")
+                .append(this.loanInterestRate).append("\n");
+
+        result.append("In total: \nAccounts: ").append(this.clientCount - 1)
+                .append(", actions: ").append(this.actionCount - 1);
+
+        return result.toString();
+    }
+
+    public ArrayList<BankClient> getAllClients() {
+        return new ArrayList<BankClient>(allClients);
+    }
+
+    public BankClient openClientAccount(String name, String surname, int age,
+                                        double balance) {
+        final var id = "MIM" + assignClientNumber();
+
+        BankClient client = new BankClient(name, surname, age, id, balance);
+
+        allClients.add(client);
+
+        return client;
+    }
+
+    public void applyForDeposit(BankClient applicant, double amountToDeposit,
+                                Timestamp toDate, double capitalizationRate) {
+
+        final var id = "DEP" + assignActionNumber();
```

Comment:
chyba sens miałoby zeby każda z akcji potrafila zwrocic swoj prefix - bylby czytelniejszy kod i bank nie musialby to tym widziec